File: packages/agent/src/infrastructure/events/types/type-guards.ts
/**
 * @fileoverview Type Guards
 *
 * Type guard functions for narrowing SessionEvent types.
 */

import type { SessionEvent } from './union.js';
import type { SessionStartEvent, SessionEndEvent, SessionForkEvent } from './session.js';
import type { UserMessageEvent, AssistantMessageEvent, SystemMessageEvent } from './message.js';
import type { ToolCallEvent, ToolResultEvent } from './tool.js';
import type { ConfigModelSwitchEvent, ConfigPromptUpdateEvent, ConfigReasoningLevelEvent } from './config.js';
import type { MessageDeletedEvent } from './message-ops.js';
import type { ContextClearedEvent } from './context.js';
import type { WorktreeAcquiredEvent, WorktreeCommitEvent, WorktreeReleasedEvent, WorktreeMergedEvent } from './worktree.js';
import type { ErrorAgentEvent, ErrorToolEvent, ErrorProviderEvent } from './error.js';
import type { RulesLoadedEvent } from './rules.js';
import type { SubagentSpawnedEvent, SubagentStatusUpdateEvent, SubagentCompletedEvent, SubagentFailedEvent } from './subagent.js';
import type { TodoWriteEvent } from './todo.js';
import type { TurnFailedEvent } from './turn.js';
import type { SkillAddedEvent, SkillRemovedEvent } from './skill.js';
import type {
  HookTriggeredEvent,
  HookCompletedEvent,
  HookBackgroundStartedEvent,
  HookBackgroundCompletedEvent,
} from './hook.js';
import type { MemoryLedgerEvent } from './memory.js';

// =============================================================================
// Type Guards
// =============================================================================

export function isSessionStartEvent(event: SessionEvent): event is SessionStartEvent {
  return event.type === 'session.start';
}

export function isSessionEndEvent(event: SessionEvent): event is SessionEndEvent {
  return event.type === 'session.end';
}

export function isSessionForkEvent(event: SessionEvent): event is SessionForkEvent {
  return event.type === 'session.fork';
}

export function isUserMessageEvent(event: SessionEvent): event is UserMessageEvent {
  return event.type === 'message.user';
}

export function isAssistantMessageEvent(event: SessionEvent): event is AssistantMessageEvent {
  return event.type === 'message.assistant';
}

export function isToolCallEvent(event: SessionEvent): event is ToolCallEvent {
  return event.type === 'tool.call';
}

export function isToolResultEvent(event: SessionEvent): event is ToolResultEvent {
  return event.type === 'tool.result';
}

export function isMessageEvent(event: SessionEvent): event is UserMessageEvent | AssistantMessageEvent | SystemMessageEvent {
  return event.type === 'message.user' || event.type === 'message.assistant' || event.type === 'message.system';
}

export function isStreamingEvent(event: SessionEvent): boolean {
  return event.type.startsWith('stream.');
}

export function isErrorEvent(event: SessionEvent): event is ErrorAgentEvent | ErrorToolEvent | ErrorProviderEvent {
  return event.type.startsWith('error.');
}

export function isWorktreeEvent(event: SessionEvent): event is WorktreeAcquiredEvent | WorktreeCommitEvent | WorktreeReleasedEvent | WorktreeMergedEvent {
  return event.type.startsWith('worktree.');
}

export function isWorktreeAcquiredEvent(event: SessionEvent): event is WorktreeAcquiredEvent {
  return event.type === 'worktree.acquired';
}

export function isWorktreeCommitEvent(event: SessionEvent): event is WorktreeCommitEvent {
  return event.type === 'worktree.commit';
}

export function isWorktreeReleasedEvent(event: SessionEvent): event is WorktreeReleasedEvent {
  return event.type === 'worktree.released';
}

export function isWorktreeMergedEvent(event: SessionEvent): event is WorktreeMergedEvent {
  return event.type === 'worktree.merged';
}

export function isRulesLoadedEvent(event: SessionEvent): event is RulesLoadedEvent {
  return event.type === 'rules.loaded';
}

export function isContextClearedEvent(event: SessionEvent): event is ContextClearedEvent {
  return event.type === 'context.cleared';
}

export function isConfigReasoningLevelEvent(event: SessionEvent): event is ConfigReasoningLevelEvent {
  return event.type === 'config.reasoning_level';
}

export function isMessageDeletedEvent(event: SessionEvent): event is MessageDeletedEvent {
  return event.type === 'message.deleted';
}

export function isConfigEvent(event: SessionEvent): event is ConfigModelSwitchEvent | ConfigPromptUpdateEvent | ConfigReasoningLevelEvent {
  return event.type.startsWith('config.');
}

export function isSubagentSpawnedEvent(event: SessionEvent): event is SubagentSpawnedEvent {
  return event.type === 'subagent.spawned';
}

export function isSubagentStatusUpdateEvent(event: SessionEvent): event is SubagentStatusUpdateEvent {
  return event.type === 'subagent.status_update';
}

export function isSubagentCompletedEvent(event: SessionEvent): event is SubagentCompletedEvent {
  return event.type === 'subagent.completed';
}

export function isSubagentFailedEvent(event: SessionEvent): event is SubagentFailedEvent {
  return event.type === 'subagent.failed';
}

export function isSubagentEvent(event: SessionEvent): event is SubagentSpawnedEvent | SubagentStatusUpdateEvent | SubagentCompletedEvent | SubagentFailedEvent {
  return event.type.startsWith('subagent.');
}

export function isTodoWriteEvent(event: SessionEvent): event is TodoWriteEvent {
  return event.type === 'todo.write';
}

export function isTodoEvent(event: SessionEvent): event is TodoWriteEvent {
  return event.type.startsWith('todo.');
}

export function isTurnFailedEvent(event: SessionEvent): event is TurnFailedEvent {
  return event.type === 'turn.failed';
}

export function isTurnEvent(event: SessionEvent): event is TurnFailedEvent {
  return event.type.startsWith('turn.');
}

export function isHookTriggeredEvent(event: SessionEvent): event is HookTriggeredEvent {
  return event.type === 'hook.triggered';
}

export function isHookCompletedEvent(event: SessionEvent): event is HookCompletedEvent {
  return event.type === 'hook.completed';
}

export function isHookBackgroundStartedEvent(event: SessionEvent): event is HookBackgroundStartedEvent {
  return event.type === 'hook.background_started';
}

export function isHookBackgroundCompletedEvent(event: SessionEvent): event is HookBackgroundCompletedEvent {
  return event.type === 'hook.background_completed';
}

export function isHookEvent(event: SessionEvent): event is HookTriggeredEvent | HookCompletedEvent | HookBackgroundStartedEvent | HookBackgroundCompletedEvent {
  return event.type.startsWith('hook.');
}

export function isSkillAddedEvent(event: SessionEvent): event is SkillAddedEvent {
  return event.type === 'skill.added';
}

export function isSkillRemovedEvent(event: SessionEvent): event is SkillRemovedEvent {
  return event.type === 'skill.removed';
}

export function isSkillEvent(event: SessionEvent): event is SkillAddedEvent | SkillRemovedEvent {
  return event.type === 'skill.added' || event.type === 'skill.removed';
}

export function isMemoryLedgerEvent(event: SessionEvent): event is MemoryLedgerEvent {
  return event.type === 'memory.ledger';
}

export function isMemoryEvent(event: SessionEvent): event is MemoryLedgerEvent {
  return event.type.startsWith('memory.');
}

File: packages/agent-rs/crates/tron-events/src/types/type_guards.rs
//! Type guard functions for [`SessionEvent`](super::base::SessionEvent).
//!
//! These replace TypeScript `isXxxEvent()` type guards. In Rust, you'd
//! normally pattern-match on `event.event_type`, but these functions are
//! convenient for filtering event collections.

use super::base::SessionEvent;
use super::event_type::EventType;

/// Whether the event is a message event (`message.user|assistant|system`).
#[must_use]
pub fn is_message_event(event: &SessionEvent) -> bool {
    event.event_type.is_message_type()
}

/// Whether the event is a streaming event (`stream.*`).
#[must_use]
pub fn is_streaming_event(event: &SessionEvent) -> bool {
    event.event_type.is_streaming_type()
}

/// Whether the event is an error event (`error.*`).
#[must_use]
pub fn is_error_event(event: &SessionEvent) -> bool {
    event.event_type.is_error_type()
}

/// Whether the event is a config event (`config.*`).
#[must_use]
pub fn is_config_event(event: &SessionEvent) -> bool {
    event.event_type.is_config_type()
}

/// Whether the event is a worktree event (`worktree.*`).
#[must_use]
pub fn is_worktree_event(event: &SessionEvent) -> bool {
    event.event_type.is_worktree_type()
}

/// Whether the event is a subagent event (`subagent.*`).
#[must_use]
pub fn is_subagent_event(event: &SessionEvent) -> bool {
    event.event_type.is_subagent_type()
}

/// Whether the event is a hook event (`hook.*`).
#[must_use]
pub fn is_hook_event(event: &SessionEvent) -> bool {
    event.event_type.is_hook_type()
}

/// Whether the event is a skill event (`skill.*`).
#[must_use]
pub fn is_skill_event(event: &SessionEvent) -> bool {
    event.event_type.is_skill_type()
}

/// Whether the event is a rules event (`rules.*`).
#[must_use]
pub fn is_rules_event(event: &SessionEvent) -> bool {
    event.event_type.is_rules_type()
}

/// Whether the event is a memory event (`memory.*`).
#[must_use]
pub fn is_memory_event(event: &SessionEvent) -> bool {
    event.event_type.is_memory_type()
}

/// Whether the event is a user message.
#[must_use]
pub fn is_user_message_event(event: &SessionEvent) -> bool {
    event.event_type == EventType::MessageUser
}

/// Whether the event is an assistant message.
#[must_use]
pub fn is_assistant_message_event(event: &SessionEvent) -> bool {
    event.event_type == EventType::MessageAssistant
}

/// Whether the event is a tool call.
#[must_use]
pub fn is_tool_call_event(event: &SessionEvent) -> bool {
    event.event_type == EventType::ToolCall
}

/// Whether the event is a tool result.
#[must_use]
pub fn is_tool_result_event(event: &SessionEvent) -> bool {
    event.event_type == EventType::ToolResult
}

/// Whether the event is a message deletion.
#[must_use]
pub fn is_message_deleted_event(event: &SessionEvent) -> bool {
    event.event_type == EventType::MessageDeleted
}

/// Whether the event is a compact boundary.
#[must_use]
pub fn is_compact_boundary_event(event: &SessionEvent) -> bool {
    event.event_type == EventType::CompactBoundary
}

/// Whether the event is a compact summary.
#[must_use]
pub fn is_compact_summary_event(event: &SessionEvent) -> bool {
    event.event_type == EventType::CompactSummary
}

/// Whether the event is context cleared.
#[must_use]
pub fn is_context_cleared_event(event: &SessionEvent) -> bool {
    event.event_type == EventType::ContextCleared
}

/// Whether the event is a session start.
#[must_use]
pub fn is_session_start_event(event: &SessionEvent) -> bool {
    event.event_type == EventType::SessionStart
}

/// Whether the event is a reasoning level config change.
#[must_use]
pub fn is_config_reasoning_level_event(event: &SessionEvent) -> bool {
    event.event_type == EventType::ConfigReasoningLevel
}

/// Whether the event is a prompt update config change.
#[must_use]
pub fn is_config_prompt_update_event(event: &SessionEvent) -> bool {
    event.event_type == EventType::ConfigPromptUpdate
}
